File: test/test_ccallback.c
typedef long rtype;
typedef rtype (*callback_t)(int);

// struct used to test storing callback
typedef struct funcs
{ int (*cb)(int);
  int	value;
} funcs;


// test callback function
rtype
ctwice(int in) {
    return in*2;
}

// default callback function
rtype
default_callback(int in) {
    return in*3;
}


// to test pass C callback from prolog
rtype
test_passcallback(callback_t func, int i)
{ return func == 0 ? default_callback(i):(func)(i);
}

// to test C callback in struct
int
test_instruct(funcs *f)
{ return (*f->cb)(f->value);
}

File: test/test_ccallback.pl
:- module(test_ccallback,
          [ test_ccallback/0
          ]).
:- use_module(library(plunit)).
:- use_module('../prolog/ffi').

/** <module> Test c callback handling

This test demonstrates c callback handling,
 both   passing  a c callback  to a
function and storing one in a struct field.
*/

test_ccallback :-
    run_tests([ ccallback ]).

c_define(mycallback, 'C'(ctwice(int, [int])) ).
:- c_import("#include \"test_ccallback.c\"",
            [ test_ccallback ],
            [ test_passcallback(mycallback, int, [int]),
              test_passcallback(null, int, [int]) as test_passcallbacknull,
              test_instruct(struct(funcs),  [int])
            ]).

do_expansion_test(R) :-
    I = 4,
    c_alloc(FPtr, struct(funcs)),
    c_store(FPtr[cb], 'C'(sym(ctwice))),
    c_store(FPtr[value], I),
    test_instruct(FPtr, R).

:- begin_tests(ccallback).

test(funcasparam, R == 4) :-
    test_passcallback(2, R).

test(struct_func, R =:= 8) :-
    % workaround to test term_expansion
    do_expansion_test(R).

test(null_callback, R == 9) :-
    test_passcallbacknull(3, R).



:- end_tests(ccallback).
